src/pymor/algorithms/dmd.py

```diff
@@ -0,0 +1,270 @@
+
+    # XL = x0, ..., xm-1  ; XR = x1, ..., xm
+    if A is not None:
+        XL = A[:-1]
+        XR = A[1:]
+
+    rank = len(XL) if target_rank is None else target_rank
+    svd = qr_svd if svd_method == 'qr_svd' else method_of_snapshots
+
+    print('SVD of XL...')
+    U, SVALS, Vh = svd(XL, product=None, modes=rank)
+
+    # Invert the Singular Values
+    SVALS_diag = np.diag(SVALS)
+    Sigma_inv = np.linalg.inv(SVALS_diag)
+    V = Vh.conj().T
+
+    # Solve the least Sqaures Problem
+    # real: A_tilde = U.T * XR * Vh.T * Sigma_inv
+    # complex: A_tilde = U.H * XR * Vh.H * Sigma_inv
+    # A = XR[:k] @ V @ Sigma_inv.inner(Uh)
+    A_tilde = U.inner(XR) @ V @ Sigma_inv
+
+    print('Calculating eigenvalue dec. ...')
+    EVALS, EVECS = spla.eig(A_tilde, b=None, left=False, right=True)
+    # omega = np.log(EVALS) / dt
+
+    omega = EVALS / dt
+
+    if order:
+        print('Sorting...')
+        # return ordered result
+        sort_idx = np.argsort(np.abs(omega))[::-1]
+        EVECS = EVECS[:, sort_idx]
+        EVALS = EVALS[sort_idx]
+        omega = omega[sort_idx]
+
+    print('Reconstructing Eigenvectors...')
+    if modes == 'standard':
+        Wk = U.lincomb(EVECS.T)
+    elif modes == 'exact' or 'exact_scaled':
+        Wk = XR.lincomb(V.T).lincomb(Sigma_inv.T).lincomb(EVECS.T)
+        if modes == 'exact_scaled':
+            EVALS_inv = np.reciprocal(EVALS)
+            Wk = Wk*EVALS_inv
+
+    return Wk, omega
+
+
+def rand_QB(A, target_rank=None, distribution='normal', oversampling=0, power_iterations=0):
+    """
+    Ramdomized QB decomposition
+
+    See Algorithm 3.1 in :cite:`EMKB19`.
+
+    Parameters
+    ----------
+    A  :
+        The |VectorArray| for which the randomized QB decomposition is to be computed.
+
+    target_rank  :  int
+        The desired rank for the decomposition. If `None`, rank = len(A).
+
+    distribution : str
+        Distribution used for the random sample matrix (`'normal'` or `'uniform'`).
+
+    oversampling : int
+        Oversampling parameter. Number of extra columns of the sample matrix.
+
+    power_iterations : int
+        Number of power iterations.
+
+
+    Returns
+    -------
+    Q :
+        |VectorArray| containig an approximate basis for the image of the input matrix A.
+        len(Q) = target_rank
+    B :
+        Numpy array. Projection of the input matrix into the lower dimensional subspace.
+    """
+    assert isinstance(A, VectorArray)
+    assert target_rank is None or target_rank <= len(A)
+    assert distribution in ('normal', 'uniform')
+
+    if A.dim == 0 or len(A) == 0:
+        return A.space.zeros(), np.zeros((target_rank, len(A)))
+
+    rank = len(A) if target_rank is None else target_rank + oversampling
+    target_rank = len(A) if target_rank is None else target_rank
+
+    Omega = np.random.normal(0, 1, (rank, len(A))) if distribution == 'normal' else np.random.rand(rank, len(A))
+
+    Y = A.lincomb(Omega)[:target_rank]
```

Comment:
This doesn't look quite right! With the slicing you just throw away all additional vectors you get from oversampling!? I think there is actually an issue with the original paper. Maybe you should look at this point more closely and also take a look at what is done in HMT11. We can also discuss this further in more detail if you wish.